### ISSUE TITLE
Memory Leak with latest neovim

### ISSUE THREAD

**Issue by cseickel** [nvim-treesitter/nvim-treesitter#2918]
### Describe the bug

There is a memory leak caused by using treesitter on the nvim 0.8 series which can be demonstrated by entering insert mode and holding down the enter key. The memory will grow quickly and never be released. This is on the latest versions, but not on a very similar setup running older versions. Both environments tested are docker images built from the same dockerfile on different dates: https://github.com/cseickel/dotfiles/blob/main/Dockerfile

I'm not sure whether this is a treesitter issue or a neovim issue, but since my only method of recreating it is with treesitter I'll start here.

Here are the versions that have the leak:
```
NVIM v0.8.0-dev+149-gcd8edb551
```
Treesitter commit: a4f9539a0c04836715c56de7e3c548bcde52b1dd

The version without the leak is running:
```
NVIM v0.7.0-dev+1434-gf89ca7194
```
and was tested with 755bfaf and a4f9539

One interesting point is that the memory grows at a similar rate on both nvim versions but it is never released on the latest. It does release very quickly on the older 0.7 (I think that was a pre-release). Both situations are problematic because the rate of increase on both is unusual, but obviously it's less of an issue on the older version where the memory is released quickly.

I tested this with lua and typescript. The size of the leak seems to correspond to the complexity of the file.

When I discovered that a leak existed, I isloated what was causing it by starting with a blank config and adding in pieces one at a time. Here are my config files: https://github.com/cseickel/dotfiles/tree/memory-leak/config/nvim

and here is what plugins are active:
https://github.com/cseickel/dotfiles/blob/memory-leak/config/nvim/lua/plugins/init.lua

### To Reproduce

1. Open a .lua file
2. `G`o to the end
3. Hit `A` to append
4. Hold down the `<enter>` key
5. Memory should climb quickly, around a MB/second on a file that is 300 lines long

### Expected behavior

Memory usage should not increase as much as it does. Even if it does, it should be released when the changes are undone.

### Output of `:checkhealth nvim-treesitter`

```markdown
vim.treesitter: require("vim.treesitter.health").check()
  ========================================================================
    - INFO: Runtime ABI version : 14
    - OK: Loaded parser for bash: ABI version 13
    - OK: Loaded parser for c_sharp: ABI version 13
    - OK: Loaded parser for css: ABI version 13
    - OK: Loaded parser for dockerfile: ABI version 13
    - OK: Loaded parser for go: ABI version 13
    - OK: Loaded parser for graphql: ABI version 13
    - OK: Loaded parser for java: ABI version 13
    - OK: Loaded parser for javascript: ABI version 13
    - OK: Loaded parser for json: ABI version 13
    - OK: Loaded parser for json5: ABI version 13
    - OK: Loaded parser for jsonc: ABI version 13
    - OK: Loaded parser for lua: ABI version 13
    - OK: Loaded parser for python: ABI version 13
    - OK: Loaded parser for r: ABI version 13
    - OK: Loaded parser for regex: ABI version 13
    - OK: Loaded parser for scss: ABI version 13
    - OK: Loaded parser for typescript: ABI version 13
    - OK: Loaded parser for vim: ABI version 13
    - OK: Loaded parser for yaml: ABI version 13
    - OK: Loaded parser for c: ABI version 13
```

### Output of `nvim --version`

```text
NVIM v0.8.0-dev+149-gcd8edb551
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-11 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=malloc -Wsuggest-attribute=cold -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az243-870

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/share/nvim"

Run :checkhealth for more info
```

### Additional context

Treesitter integration is awesome, thanks for all of your hard work!

**Comment by cseickel** [comment]
I just noticed that my `checkhealth` shows a different runtime ABI version from the parsers. I can tell you that it is the same on the computer without the leak. Is it a problem if treesitter is newer than the parsers?

**Comment by clason** [comment]
First, it sounds like this is an issue more with neovim's tree-sitter integration than nvim-treesitter. It would be helpful if you could

0. revert nvim-treesitter before the recent refactors for 0.7+ support (and test if it still happens, just to make sure -- but this step is really only to set the stage).
1. bisect Neovim to try to find when the regression occurred (either manually -- check for `feat(treesitter)` commits -- or using `git bisect`).
2. if the offending commit was a tree-sitter bump, bisect tree-sitter itself by changing the release URL to  a specific commit).

Also, there's a bunch of seemingly unrelated plugins in your "minimal" config -- are you sure that these are strictly necessary to reproduce the issue? If so, that would make this much more complicated to track down.

**Comment by cseickel** [comment]
Thanks for the quick response. I'll go through your list to try to nail down where the issue occurred. I love that it starts with zero!

My config is certainly not the "minimal" config needed, but it is "a" config that causes the issue. I'll work on a more useful one for testing.

My apologies for rushing an incomplete issue, but I wanted to get it out there sooner rather than later in case there were others experiencing the same thing and looking for clues.

**Comment by cseickel** [comment]
As it turns out, the difference is NOT the versions being used. The computer with the leak has the leak no matter what version of neovim I run, including with the exact same commit as the computer with the problem. I haven't been able to figure out what the difference is but I am still looking.

Here is the actual minimal init needed to recreate the issue:
```vim
packadd nvim-treesitter
lua << EOF
  require'nvim-treesitter.configs'.setup {
    ensure_installed = {
      "lua",
    },
    highlight = {
      enable = true,
    },
    indent = {
      enable = true
    }
  }
EOF
```

and I run that with:
```
nvim --noplugins -u ~/memory-leak-test.vim
```

Then open any `.lua` file, go to the end, enter insert mode, and hold down the enter key. I have been using this file for testing but it will work with anything: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/v2.x/lua/neo-tree/utils.lua

With the above file, memory usage will climb from 20 mb to 90 mb by the time I get to line 1000.

Does anyone else experience the same issue as me?

**Comment by ViRu-ThE-ViRuS** [comment]
I see similar numbers as you mention with the minimal config, but its still noticeable growth in memory usage based on just seeing it in Activity Monitor on Os.

In a C file (with appropriate addition to minimal config), it grows from 20->100+ within 2000 new lines on top of 1000 existing ones.

In a lua file, grows from 20->steadily onwards, ~1mb/100 lines, though this file is originally like 30 lines so that might be a factor here in terms of file complexity.

**Comment by vigoux** [comment]
Might be related: https://github.com/neovim/neovim/issues/14216

**Comment by silvercircle** [comment]
The issue is still present in current Neovim:

    NVIM v0.9.0-dev-565+gd215dae0e
    Build type: Release
    LuaJIT 2.1.0-beta3

```
    indent = {
      enable = true,
    },
```
in the treesitter config is probably the cause.  Setting it to false cures the memory leak for me.

checkhealth output:
```
Parser/Features         H L F I J
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - yaml                ✓ ✓ ✓ ✓ ✓
  - go                  ✓ ✓ ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - objc                ✓ . . . .
  - c_sharp             ✓ ✓ ✓ . ✓
  - ada                 ✓ ✓ ✓ . .
  - html                ✓ ✓ ✓ ✓ ✓
  - scala               ✓ . ✓ . ✓
  - rust                ✓ ✓ ✓ ✓ ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - dart                ✓ ✓ . ✓ ✓
  - org                 ✓ . . . ✓   
```
Basically all parsers that support indent are affected. Others (like Ada or Scala) do not show the leak

**Comment by onemoresuza** [comment]
Tested on a `beancount` file greater than **140Kb** with only `highlight` enabled and the leakage still occurred.

Here's the `nix flake` used for testing, which uses `Neovim nightly` with `nvim-treesitter` as the sole installed plugin and `beancount` as the only parser, while just the `highlight` module is enabled:

``` nix
{
  description = "A Neovim setup for Treesitter memory leakege testing";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs";
    flake-utils.url = "github:numtide/flake-utils";
    neovim-nightly-overlay.url = "github:nix-community/neovim-nightly-overlay";
  };

  outputs = { self, nixpkgs, flake-utils, neovim-nightly-overlay }:
    flake-utils.lib.eachDefaultSystem
        (system:
            let
                pkgs = import nixpkgs {
                    inherit system;
                    overlays = [
                        neovim-nightly-overlay.overlay
                    ];
                };

                bootstrapedNeovim = with pkgs; neovim.override {
                    configure = {
                        packages.memoryLeakgeTest = with vimPlugins; {
                            start = [
                                (nvim-treesitter.withPlugins (
                                    plugins: with plugins; [ beancount ]
                                ))
                            ];
                        };
                        customRC = ''
                            lua << EOF
                                require("nvim-treesitter.configs").setup({
                                    highlight = { enable = true },
                                })
                            EOF
                        '';
                    };
                };
            in {
                devShells.default = pkgs.mkShell {
                    name = "nvim-memory-leakege";

                    buildInputs = with pkgs; [
                        bootstrapedNeovim
                    ];

                    #
                    # Avoid messing with an already installed Neovim setup.
                    #
                    shellHook = ''
                        export HOME="$PWD/test-home"
                        mkdir -p "$HOME"
                        export XDG_DATA_HOME="$HOME/xdg-data"
                        mkdir -p "$XDG_DATA_HOME"
                        export XDG_CONFIG_HOME="$HOME/xdg-config"
                        mkdir -p "$XDG_CONFIG_HOME"
                        export XDG_STATE_HOME="$HOME/xdg-state"
                        mkdir -p "$XDG_STATE_HOME"
                    '';
                };
            }
        );
}
```
And its `lock file`:
``` json
{
  "nodes": {
    "flake-compat": {
      "flake": false,
      "locked": {
        "lastModified": 1673956053,
        "narHash": "sha256-4gtG9iQuiKITOjNQQeQIpoIB6b16fm+504Ch3sNKLd8=",
        "owner": "edolstra",
        "repo": "flake-compat",
        "rev": "35bb57c0c8d8b62bbfd284272c928ceb64ddbde9",
        "type": "github"
      },
      "original": {
        "owner": "edolstra",
        "repo": "flake-compat",
        "type": "github"
      }
    },
    "flake-utils": {
      "locked": {
        "lastModified": 1678901627,
        "narHash": "sha256-U02riOqrKKzwjsxc/400XnElV+UtPUQWpANPlyazjH0=",
        "owner": "numtide",
        "repo": "flake-utils",
        "rev": "93a2b84fc4b70d9e089d029deacc3583435c2ed6",
        "type": "github"
      },
      "original": {
        "owner": "numtide",
        "repo": "flake-utils",
        "type": "github"
      }
    },
    "flake-utils_2": {
      "locked": {
        "lastModified": 1667395993,
        "narHash": "sha256-nuEHfE/LcWyuSWnS8t12N1wc105Qtau+/OdUAjtQ0rA=",
        "owner": "numtide",
        "repo": "flake-utils",
        "rev": "5aed5285a952e0b949eb3ba02c12fa4fcfef535f",
        "type": "github"
      },
      "original": {
        "owner": "numtide",
        "repo": "flake-utils",
        "type": "github"
      }
    },
    "neovim-flake": {
      "inputs": {
        "flake-utils": "flake-utils_2",
        "nixpkgs": [
          "neovim-nightly-overlay",
          "nixpkgs"
        ]
      },
      "locked": {
        "dir": "contrib",
        "lastModified": 1680569951,
        "narHash": "sha256-9JRp6c7Iybr13jH1ZnBIiglfUZd2sRnys+M7E6o9SmU=",
        "owner": "neovim",
        "repo": "neovim",
        "rev": "7c8c1550737b34fabb3e4cecbc6b6bf3581b2235",
        "type": "github"
      },
      "original": {
        "dir": "contrib",
        "owner": "neovim",
        "repo": "neovim",
        "type": "github"
      }
    },
    "neovim-nightly-overlay": {
      "inputs": {
        "flake-compat": "flake-compat",
        "neovim-flake": "neovim-flake",
        "nixpkgs": "nixpkgs"
      },
      "locked": {
        "lastModified": 1680596024,
        "narHash": "sha256-tskY93I+q1XuxU5CWriWZhnHVGSP08WRoHCed01n2Sw=",
        "owner": "nix-community",
        "repo": "neovim-nightly-overlay",
        "rev": "7070f17bb65146f9f6cff012c0321cbc9c8c8def",
        "type": "github"
      },
      "original": {
        "owner": "nix-community",
        "repo": "neovim-nightly-overlay",
        "type": "github"
      }
    },
    "nixpkgs": {
      "locked": {
        "lastModified": 1680577567,
        "narHash": "sha256-Y4W57i0TzczqwMTjfVSbAZT5RJx7u+mpQIW/ofrkTQs=",
        "owner": "nixos",
        "repo": "nixpkgs",
        "rev": "b01f185e4866de7c5b5a82f833ca9ea3c3f72fc4",
        "type": "github"
      },
      "original": {
        "owner": "nixos",
        "ref": "nixpkgs-unstable",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "nixpkgs_2": {
      "locked": {
        "lastModified": 1680655309,
        "narHash": "sha256-99I+4RD4NXaNNi4/u8YtkVFR2vk4XS06qBWAQWz5mlg=",
        "owner": "nixos",
        "repo": "nixpkgs",
        "rev": "7884a01ebfed4b4ee8abc39113d1417665b9b05a",
        "type": "github"
      },
      "original": {
        "owner": "nixos",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "flake-utils": "flake-utils",
        "neovim-nightly-overlay": "neovim-nightly-overlay",
        "nixpkgs": "nixpkgs_2"
      }
    }
  },
  "root": "root",
  "version": 7
}
```

**Comment by lewis6991** [comment]
I've been able to repro this and I'm fairly confident this is due to `indentexpr` and `memoize_by_buftick()` since the cache is not properly managed: https://github.com/nvim-treesitter/nvim-treesitter/blob/26230f06b6ee5a456c6114860ba792382922fadf/lua/nvim-treesitter/ts_utils.lua#L354

Making the cache a fully weak table fixes the issue for me:

```diff
diff --git a/lua/nvim-treesitter/ts_utils.lua b/lua/nvim-treesitter/ts_utils.lua
index 373bae34..66225007 100644
--- a/lua/nvim-treesitter/ts_utils.lua
+++ b/lua/nvim-treesitter/ts_utils.lua
@@ -351,7 +351,7 @@ function M.memoize_by_buf_tick(fn, options)
   options = options or {}

   ---@type table<string, {result: any, last_tick: integer}>
-  local cache = {}
+  local cache = setmetatable({}, { __mode = 'kv' }) ---@type table<any,any>
   local bufnr_fn = utils.to_func(options.bufnr or utils.identity)
   local key_fn = utils.to_func(options.key or utils.identity)
```

**Comment by lewis6991** [comment]
Should be fixed in #4812
